Content.Server/Chemistry/EntitySystems/ChemMasterSystem.cs

```diff
@@ -1,20 +1,239 @@
+
+            if (fromBuffer)
+                bufferSolution.RemoveReagent(reagentId, amount);
+            else if (containerSlot is not null &&
+                containerSlot.HasItem &&
+                containerSlot.Item is {Valid: true} container &&
+                TryComp(container, out FitsInDispenserComponent? fits) &&
+                _solutionContainerSystem.TryGetSolution(container, fits.Solution, out var containerSolution))
+            {
+                _solutionContainerSystem.TryRemoveReagent(container, containerSolution, reagentId, amount);
+            }
+
+            UpdateUiState(chemMaster, updateLabel: fromBuffer);
+        }
+
+        private void OnCreatePillsMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterCreatePillsMessage message)
+        {
+            // Ensure the amount is valid.
+            if (message.Amount == 0 || message.Amount > chemMaster.PillProductionLimit)
+                return;
+
+            CreatePillsOrBottles(chemMaster, pills: true, message.Amount, message.Label, message.Session.AttachedEntity);
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void OnCreateBottlesMessage(EntityUid uid, ChemMasterComponent chemMaster, ChemMasterCreateBottlesMessage message)
+        {
+            // Ensure the amount is valid.
+            if (message.Amount == 0 || message.Amount > chemMaster.BottleProductionLimit)
+                return;
+
+            CreatePillsOrBottles(chemMaster, pills: false, message.Amount, message.Label, message.Session.AttachedEntity);
+            UpdateUiState(chemMaster);
+            ClickSound(chemMaster);
+        }
+
+        private void CreatePillsOrBottles(ChemMasterComponent chemMaster, bool pills, FixedPoint2 amount, string label, EntityUid? user)
+        {
+            var bufferSolution = _solutionContainerSystem.EnsureSolution(chemMaster.Owner, ChemMasterComponent.SolutionName);
+            var filter = user.HasValue ? Filter.Entities(user.Value) : Filter.Empty();
+            if (bufferSolution.TotalVolume == 0)
+            {
+                _popupSystem.PopupCursor(Loc.GetString("chem-master-window-buffer-empty-text"), filter);
+                return;
+            }
+
+            var individualVolume = FixedPoint2.Min(bufferSolution.TotalVolume / amount, FixedPoint2.New(pills ? 50 : 30));
+            if (individualVolume < FixedPoint2.New(1))
+            {
+                _popupSystem.PopupCursor(Loc.GetString("chem-master-window-buffer-low-text"), filter);
+                return;
+            }
+
+            for (int i = 0; i < amount; i++)
+            {
+                var item = Spawn(pills ? "Pill" : "ChemistryEmptyBottle01", Transform(chemMaster.Owner).Coordinates);
+
+                if (label != "")
+                {
+                    var labelComponent = item.EnsureComponent<LabelComponent>();
```

Comment:
```suggestion
                    var labelComponent = EnsureComp<LabelComponent>(item);
```
don't use the entityuid methods. didn't even know we still had them